Mathlib/Order/BooleanSubalgebra.lean

```diff
@@ -0,0 +1,379 @@
+
+/-- A boolean subalgebra of a lattice inherits a Heyting implication. -/
+instance instHImpCoe : HImp L where himp a b := ⟨a ⇨ b, himp_mem a.2 b.2⟩
+
+@[simp, norm_cast] lemma val_bot : (⊥ : L) = (⊥ : α) := rfl
+@[simp, norm_cast] lemma val_top : (⊤ : L) = (⊤ : α) := rfl
+@[simp, norm_cast] lemma val_sup (a b : L) : a ⊔ b = (a : α) ⊔ b := rfl
+@[simp, norm_cast] lemma val_inf (a b : L) : a ⊓ b = (a : α) ⊓ b := rfl
+@[simp, norm_cast] lemma val_compl (a : L) : aᶜ = (a : α)ᶜ := rfl
+@[simp, norm_cast] lemma val_sdiff (a b : L) : a \ b = (a : α) \ b := rfl
+@[simp, norm_cast] lemma val_himp (a b : L) : a ⇨ b = (a : α) ⇨ b := rfl
+
+@[simp] lemma mk_bot : (⟨⊥, bot_mem⟩ : L) = ⊥ := rfl
+@[simp] lemma mk_top : (⟨⊤, top_mem⟩ : L) = ⊤ := rfl
+@[simp] lemma mk_sup_mk (a b : α) (ha hb) : (⟨a, ha⟩ ⊔ ⟨b, hb⟩ : L) = ⟨a ⊔ b, L.supClosed ha hb⟩ :=
+  rfl
+@[simp] lemma mk_inf_mk (a b : α) (ha hb) : (⟨a, ha⟩ ⊓ ⟨b, hb⟩ : L) = ⟨a ⊓ b, L.infClosed ha hb⟩ :=
+  rfl
+@[simp] lemma compl_mk (a : α) (ha) : (⟨a, ha⟩ : L)ᶜ = ⟨aᶜ, compl_mem ha⟩ := rfl
+@[simp] lemma mk_sdiff_mk (a b : α) (ha hb) : (⟨a, ha⟩ \ ⟨b, hb⟩ : L) = ⟨a \ b, sdiff_mem ha hb⟩ :=
+  rfl
+@[simp] lemma mk_himp_mk (a b : α) (ha hb) : (⟨a, ha⟩ ⇨ ⟨b, hb⟩ : L) = ⟨a ⇨ b, himp_mem ha hb⟩ :=
+  rfl
+
+/-- A boolean subalgebra of a lattice inherits a boolean algebra structure. -/
+instance instBooleanAlgebraCoe (L : BooleanSubalgebra α) : BooleanAlgebra L :=
+  Subtype.coe_injective.booleanAlgebra _ val_sup val_inf val_top val_bot val_compl val_sdiff
+    val_himp
+
+/-- The natural lattice hom from a boolean subalgebra to the original lattice. -/
+def subtype (L : BooleanSubalgebra α) : BoundedLatticeHom L α where
+  toFun := ((↑) : L → α)
+  map_bot' := L.val_bot
+  map_top' := L.val_top
+  map_sup' := val_sup
+  map_inf' := val_inf
+
+@[simp, norm_cast] lemma coe_subtype (L : BooleanSubalgebra α) : L.subtype = ((↑) : L → α) := rfl
+lemma subtype_apply (L : BooleanSubalgebra α) (a : L) : L.subtype a = a := rfl
+
+lemma subtype_injective (L : BooleanSubalgebra α) : Injective <| subtype L := Subtype.coe_injective
+
+/-- The inclusion homomorphism from a boolean subalgebra `L` to a bigger boolean subalgebra `M`. -/
+def inclusion (h : L ≤ M) : BoundedLatticeHom L M where
+  toFun := Set.inclusion h
+  map_bot' := rfl
+  map_top' := rfl
+  map_sup' _ _ := rfl
+  map_inf' _ _ := rfl
+
+@[simp] lemma coe_inclusion (h : L ≤ M) : inclusion h = Set.inclusion h := rfl
+lemma inclusion_apply (h : L ≤ M) (a : L) : inclusion h a = Set.inclusion h a := rfl
+
+lemma inclusion_injective (h : L ≤ M) : Injective <| inclusion h := Set.inclusion_injective h
+
+@[simp] lemma inclusion_rfl (L : BooleanSubalgebra α) : inclusion le_rfl = .id L := rfl
+@[simp] lemma subtype_comp_inclusion (h : L ≤ M) : M.subtype.comp (inclusion h) = L.subtype := rfl
+
+/-- The maximum boolean subalgebra of a lattice. -/
+instance instTop : Top (BooleanSubalgebra α) where
```

Comment:
Let's have some simp lemmas describing membership of this subalgebra, and similarly for bot, inf, and infset. Similarly for sup and supset, after they're defined.

Reply:
I can't see the `sup` and `supSet` versions?

Reply:
I will let you tell me what they should say 😉

(hint: infimum of subobjects typically corresponds to intersection of the underlying sets, but supremum does not)

Reply:
Ah, I remember we discussed this now, and I misremembered the conclusion :) You said that the infimum and supremum don't correspond to intersection and union here, and I said that you were wrong. But you were only wrong about infimum!

Reply:
I say "not (inf = inter and sup = union)", you hear "not inf = inter and not sup = union" 😝

Reply:
Actually you said "Coercion to sets doesn't preserve suprema nor infima", I find it hard to believe "not inf = inter and not sup = union" is the wrong interpretation :D

Reply:
Ah yes, it does not preserve _arbitrary_ suprema/infima, since the bottom boolean subalgebra is not the empty set. This is probably what I had in mind when I said this

Reply:
Uh, isn't that exactly what `coe_iInf` says? It's also what you agreed was true when we talked about this previously!

Reply:
Okay, at this point I don't remember what I said and why